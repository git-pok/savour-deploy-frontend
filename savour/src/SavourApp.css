.SavourApp {
  display: block;
  text-align: center;
  width: 100%;
  height: 100%;
}