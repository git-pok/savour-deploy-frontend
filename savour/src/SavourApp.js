import './SavourApp.css';
import { Redirect, useHistory } from 'react-router-dom';
import { useContext, useCallback } from 'react';
import Navbar from './Navbar.js';
import Routes from './Routes.js';
import useLocalStorage from './hooks/useLocalStorage.js';
import UserContext from './context/UserContext.js';
import useToggleState from './hooks/useToggleState';
import SavourApi from './models/SavourApi.js';

/**
 * SavourApp
 * Props: none
 * Renders savour app with navbar, routes, and context.
*/
const SavourApp = () => {
  const [ isLoggedOut, setIsLoggedOut ] = useToggleState(false);
  const [ usrData, setUsrData ] = useLocalStorage("userData", null);
  const history = useHistory();

  const logOut = useCallback(() => {
    window.localStorage.clear();
    setUsrData(() => null);
    SavourApi.token = null;
    history.push("/login");
  }, [usrData]);

  const userLink = usrData ? `${usrData.userUsername}` : "";
  const linkNames = [
    ["recipes", "/recipes"], ["shoplists", "/shoppinglists"],
    ["recipelists", "/recipelists"], ["favs", "/favs"],
    ["saved", "/saved"], ["profile", `/users/${userLink}`],
    ["contact", "/contact"]
  ];

  return (
    <div className="SavourApp">
      <UserContext.Provider value={{usrData, setUsrData, logOut}}>
        <Navbar
          linkNames={linkNames}
          logOut={logOut} />
        <Routes />
      </UserContext.Provider>
    </div>
  );
}

export default SavourApp;
