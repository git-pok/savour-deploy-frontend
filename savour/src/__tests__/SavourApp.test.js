import SavourApp from '../SavourApp.js';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';


afterAll(() => {
  const { getByText } = render(
    <MemoryRouter intitialEntries={["/"]}>
      <SavourApp />
    </MemoryRouter>
  );
  const logoutBtn = getByText("LOG OUT");
  fireEvent.click(logoutBtn);
});


function renderSavourApp () {
  return render(
  <MemoryRouter intitialEntries={["/"]}>
    <SavourApp />
  </MemoryRouter>
  );
}

test("renders SavourApp without crashing", () => {
  render(<MemoryRouter><SavourApp /></MemoryRouter>);
});

test("renders snapshot without crashing", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <SavourApp />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

test("login form submission", async () => {
  const { findByText, getByText, getByTestId, queryByLabelText } = render(
        <MemoryRouter>
          <SavourApp />
        </MemoryRouter>
  );

  const loginBtn = getByText("LOG IN");
  fireEvent.click(loginBtn);
  const submitBtn = getByTestId("sbmtBtn");
  const signInUsrname = queryByLabelText("Username");
  const signInPassword = queryByLabelText("Password");
  fireEvent.change(signInUsrname, { target: { value: "testUsr" } });
  fireEvent.change(signInPassword, { target: { value: "password" } });
  fireEvent.click(submitBtn);
  const welcomeHdr = await findByText("Welcome to Savour – The recipe app");
  expect(welcomeHdr).toBeInTheDocument();
  const welcomeUsrText = await findByText("WELCOME TESTUSR!");
  expect(welcomeUsrText).toBeInTheDocument();
});

test("home page to shoppinglists page", async () => {
  const { findByText, getByText } = renderSavourApp();

  const welcome = getByText("Welcome to Savour – The recipe app");
  expect(welcome).toBeInTheDocument();
  const shplistsBtn = getByText("SHOPLISTS");
  fireEvent.click(shplistsBtn);
  const shplistsText = await findByText("testUsr's Shoppinglists");
  expect(shplistsText).toBeInTheDocument();
});

test("profile page to contact page", async () => {
  const { findByText, getByText } = renderSavourApp();

  const welcome = getByText("Welcome to Savour – The recipe app");
  expect(welcome).toBeInTheDocument();
  const profileBtn = getByText("PROFILE");
  fireEvent.click(profileBtn);
  const profileText = await findByText("testUsr Details");
  expect(profileText).toBeInTheDocument();
  const contactBtn = getByText("CONTACT");
  fireEvent.click(contactBtn);
  const contactText = await findByText("App Creator and Developer Information");
  expect(contactText).toBeInTheDocument();
});