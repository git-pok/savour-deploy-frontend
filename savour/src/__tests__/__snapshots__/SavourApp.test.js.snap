// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot without crashing 1`] = `
<DocumentFragment>
  <div
    class="SavourApp"
  >
    <div
      class="Navbar-mobile-menu"
    >
      <p>
        NAV MENU
      </p>
    </div>
    <nav
      class="Navbar"
    >
      <div
        class="Navbar-left-links"
      >
        <a
          href="/"
        >
          SAVOUR
        </a>
      </div>
      <div
        class="Navbar-right-links"
      >
        <a
          href="/login"
        >
          LOG IN
        </a>
        <a
          href="/signup"
        >
          SIGN UP
        </a>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
